File: Codeforces/Intel 5 oct/C.cpp
#include <bits/stdc++.h>
using namespace std;

#define sld(a) scanf("%lld",&a)
#define pld(a) printf("%lld\n",a) 
#define ll long long
#define min1 -100000000000099

ll yo(ll val){
	if(val > 0){
		return val;
	}
	return 0;
}

struct seg{
	ll suff,pref,tot,seg1;
};

seg st[512345];
ll lis[112345];

seg comb(seg a,seg b){
	seg c;
	c.pref = max(a.pref,a.tot+b.pref);
	c.suff = max(b.suff,a.suff+b.tot);
	c.tot = a.tot+ b.tot;
	c.seg1 = max(a.seg1,max(b.seg1,a.suff+b.pref));
	return c;
}

void cons(ll start,ll end,ll node){
	if(start == end){
		st[node].suff = st[node].pref = st[node].tot = st[node].seg1 = lis[start];
		return;
	}
	ll mid = (start+end)/2;
	ll le = 2*node+1;
	ll ri = 2*node+2;
	cons(start,mid,le);
	cons(mid+1,end,ri);
	st[node] = comb(st[le],st[ri]);
	return;
}

void update(ll start,ll end,ll node,ll pos){
	if(start == end){
		st[node].suff = st[node].pref = st[node].tot = st[node].seg1 = min1;
		return ;
	}
	ll mid = (start+end)/2;
	ll le = 2*node+1;
	ll ri = 2*node+2;
	if(pos <= mid){
		update(start,mid,le,pos);
	}
	else{
		update(mid+1,end,ri,pos);
	}
	st[node] = comb(st[le],st[ri]);
	return ;
}

int main(){
	ll n;
	sld(n);
	int order[n];
	for(ll i=0;i<n;i++){
		sld(lis[i]);
	}
	for(int j=0;j<n;j++){
		scanf("%d",&order[j]);
	}
	cons(0,n-1,0);
	for(int i=0;i<n;i++){
		update(0,n-1,0,order[i]-1);
		pld(yo(st[0].seg1));
	}
	return 0;
}

File: Codeforces/Intel 5 oct/C.py
class seg:
	def __init__(self,pre,suff,tot,seg1):
		self.suff = suff
		self.pre = pre
		self.tot = tot
		self.seg1 = seg1

a = seg(0,0,0,0)
st = [a]*(5*10**5)

def comb(a,b):
	c = seg(0,0,0,0)
	c.pre = max(a.pre,a.tot + b.pre)
	c.suff = max(b.suff,a.suff+b.tot)
	c.tot = a.tot + b.tot
	c.seg1 = max(a.seg1,b.seg1,a.suff+b.pre)
	return c

def build(start,end,node):
	if start == end:
		yo = lis[start]
		st[node] = seg(yo,yo,yo,yo)
		return
	le = 2*node+1
	ri = 2*node+2
	mid = (start+end)/2
	build(start,mid,le)
	build(mid+1,end,ri)
	st[node] = comb(st[le],st[ri])


def update(start,end,node,pos):
	if start == end:
		yo = -(10**15)
		st[node] = seg(yo,yo,yo,yo)
		return
	mid = (start+end)/2
	le = 2*node+1
	ri = 2*node+2
	if pos <= mid:
		update(start,mid,le,pos)
	else:
		update(mid+1,end,ri,pos)
	
	st[node] = comb(st[le],st[ri])

n = input()
lis = map(int,raw_input().split())
order = map(int,raw_input().split())

build(0,n-1,0)
for ele in order:
	update(0,n-1,0,ele-1)
	print max(0,st[0].seg1)